File: google-oauth/python/auth/dummy.py
"""
Dummy authentication for testing.

Only active when ``enable_dummy_auth`` is ``True`` in the application
settings. Never enable this in production -- it bypasses all authentication.

This module contains only dummy auth logic. Route registration is handled
in ``main.py``, keeping framework wiring separate from authentication logic.
"""

import json
import logging
import uuid
from pathlib import Path

from fastapi import Request
from fastapi.responses import HTMLResponse, RedirectResponse, Response
from fastapi.templating import Jinja2Templates

from config import Settings
from models import Session, User
from session.middleware import set_session_cookie
from session.store import AbstractSessionStore

__all__ = ["DummyAuth"]

logger = logging.getLogger(__name__)


def _load_users_from_file(users_file: Path) -> dict[str, User]:
    """
    Loads dummy users from a JSON file.

    Returns an empty dict (with a warning) if the file is missing or malformed,
    so the server still starts cleanly without dummy users available.
    """
    try:
        raw_entries: list[dict[str, str]] = json.loads(users_file.read_text(encoding="utf-8"))
        return {entry["id"]: User(**entry) for entry in raw_entries}
    except (OSError, ValueError, KeyError):
        logger.warning("Could not load dummy users from %s -- dummy login unavailable.", users_file)
        return {}


class DummyAuth:
    """
    Provides instant sign-in as a pre-defined test user.

    Exposes two async handler methods -- ``handle_login`` and ``handle_list``
    -- which are registered as routes in ``main.py``. This class contains only
    dummy auth logic; it has no knowledge of URL paths or FastAPI's routing API.

    Users are loaded from the file path given in ``settings.dummy_users_file``.
    """

    def __init__(
        self,
        store: AbstractSessionStore,
        settings: Settings,
        templates: Jinja2Templates,
    ) -> None:
        self._store: AbstractSessionStore = store
        self._settings: Settings = settings
        self._enabled: bool = settings.enable_dummy_auth
        self._templates: Jinja2Templates = templates
        self._users: dict[str, User] = (
            _load_users_from_file(settings.dummy_users_file) if self._enabled else {}
        )
        if self._enabled:
            logger.info(
                "Dummy auth enabled -- %d user(s) available: %s.",
                len(self._users),
                ", ".join(self._users),
            )

    @property
    def enabled(self) -> bool:
        """True when dummy auth is active."""
        return self._enabled

    async def handle_login(self, request: Request) -> Response:
        """Signs in immediately as the test user matching ``?id=<id>``."""
        if not self._enabled:
            return HTMLResponse(
                "Dummy auth is not enabled. Set ENABLE_DUMMY_AUTH=true.",
                status_code=403,
            )

        user_id: str = request.query_params.get("id", "")
        user: User | None = self._users.get(user_id)

        if user is None:
            available_ids: str = ", ".join(self._users)
            return HTMLResponse(
                f"Unknown test user id \"{user_id}\". Available: {available_ids}",
                status_code=404,
            )

        session: Session = Session(id=str(uuid.uuid4()), user=user)
        self._store.save(session)
        logger.info("Dummy sign-in as %s (%s).", user.name, user.id)

        redirect_response: RedirectResponse = RedirectResponse("/", status_code=302)
        set_session_cookie(redirect_response, session.id, self._settings)
        return redirect_response

    async def handle_list(self, request: Request) -> Response:
        """Returns an HTML page listing all available test users."""
        if not self._enabled:
            return HTMLResponse("Dummy auth is not enabled.", status_code=403)

        return self._templates.TemplateResponse(
            request,
            "dummy_list.html",
            {"users": list(self._users.values())},
        )

File: google-oauth/typescript/src/auth/dummy.ts
import fs from "fs";
import path from "path";
import { v4 as uuidv4 } from "uuid";
import type { AppRequest, AppResponse, User } from "../types.js";
import type { SessionStore } from "../session/store.js";
import { setSessionCookie } from "../session/middleware.js";

const DUMMY_FILE = path.resolve("../dummy_users.json");

/**
 * Loads dummy users from the shared dummy_users.json file.
 * Returns an empty array (with a warning) if the file is missing or malformed.
 */
function loadDummyUsers(): Map<string, User> {
  try {
    const raw = fs.readFileSync(DUMMY_FILE, "utf8");
    const users = JSON.parse(raw) as User[];
    return new Map(users.map((u) => [u.id, u]));
  } catch {
    console.warn(`[DummyAuth] Could not load ${DUMMY_FILE} — dummy login disabled.`);
    return new Map();
  }
}

/**
 * Provides dummy sign-in for testing.
 *
 * Only active when ENABLE_DUMMY_AUTH=true in the environment.
 * Visit /auth/dummy?id=<user-id> to sign in as that user.
 * Users are defined in dummy_users.json at the project root.
 */
export class DummyAuth {
  private readonly users: Map<string, User>;
  private readonly enabled: boolean;

  constructor(private readonly store: SessionStore) {
    this.enabled = process.env.ENABLE_DUMMY_AUTH === "true";
    this.users = this.enabled ? loadDummyUsers() : new Map();
    if (this.enabled) {
      console.log(`[DummyAuth] Enabled with ${this.users.size} user(s): ${[...this.users.keys()].join(", ")}`);
    }
  }

  isEnabled(): boolean {
    return this.enabled;
  }

  /**
   * Signs in as the dummy user matching ?id=<id>.
   * Responds 404 if the id is unknown, 403 if dummy auth is disabled.
   */
  handleLogin(req: AppRequest, res: AppResponse): void {
    if (!this.enabled) {
      res.writeHead(403);
      res.end("Dummy auth is not enabled. Set ENABLE_DUMMY_AUTH=true.");
      return;
    }

    const url = new URL(req.url ?? "/", `http://${req.headers.host}`);
    const id = url.searchParams.get("id") ?? "";
    const user = this.users.get(id);

    if (!user) {
      const list = [...this.users.keys()].join(", ");
      res.writeHead(404);
      res.end(`Unknown dummy user id "${id}". Available: ${list}`);
      return;
    }

    const session = { id: uuidv4(), user, createdAt: Date.now() };
    this.store.set(session);
    setSessionCookie(res, session.id);
    console.log(`[DummyAuth] Signed in as ${user.name} (${user.id})`);
    res.writeHead(302, { Location: "/" });
    res.end();
  }

  /**
   * Returns an HTML page listing all available dummy users with sign-in links.
   */
  handleList(_req: AppRequest, res: AppResponse): void {
    if (!this.enabled) {
      res.writeHead(403);
      res.end("Dummy auth is not enabled.");
      return;
    }

    const items = [...this.users.values()]
      .map(
        (u) => `
        <li>
          <img src="${u.picture}" width="36" height="36" alt="">
          <div>
            <strong>${u.name}</strong><br>
            <small>${u.email}</small>
          </div>
          <a href="/auth/dummy?id=${u.id}">Sign in</a>
        </li>`
      )
      .join("");

    const body = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Dummy Login</title>
  <style>
    *, *::before, *::after { box-sizing: border-box; margin: 0; padding: 0; }
    body {
      font-family: system-ui, sans-serif;
      background: #f5f5f5;
      min-height: 100vh;
      display: flex;
      align-items: center;
      justify-content: center;
    }
    .card {
      background: #fff;
      border-radius: 10px;
      padding: 2rem 2.5rem;
      box-shadow: 0 2px 12px rgba(0,0,0,0.08);
      max-width: 400px;
      width: 100%;
    }
    h1 { font-size: 1.2rem; margin-bottom: 0.25rem; }
    p { color: #888; font-size: 0.85rem; margin-bottom: 1.5rem; }
    ul { list-style: none; display: flex; flex-direction: column; gap: 0.75rem; }
    li {
      display: flex;
      align-items: center;
      gap: 0.75rem;
      padding: 0.6rem 0.75rem;
      border: 1px solid #eee;
      border-radius: 8px;
    }
    li img { border-radius: 50%; flex-shrink: 0; }
    li div { flex: 1; }
    li small { color: #888; }
    li a {
      padding: 0.35rem 0.85rem;
      background: #222;
      color: #fff;
      text-decoration: none;
      border-radius: 5px;
      font-size: 0.85rem;
      white-space: nowrap;
    }
    li a:hover { background: #444; }
  </style>
</head>
<body>
  <div class="card">
    <h1>Dummy Login</h1>
    <p>Testing only. Users are defined in dummy_users.json.</p>
    <ul>${items}</ul>
  </div>
</body>
</html>`;

    res.writeHead(200, { "Content-Type": "text/html; charset=utf-8" });
    res.end(body);
  }
}
